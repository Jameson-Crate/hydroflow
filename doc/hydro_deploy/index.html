<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `hydro_deploy` crate."><title>hydro_deploy - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Bold-a2c9cd1067f8b328.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-9ee3a5e31a2afa3e.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="hydro_deploy" data-themes="" data-resource-suffix="" data-rustdoc-version="1.75.0-nightly (cd674d617 2023-10-24)" data-channel="nightly" data-search-js="search-8fbf244ebcf71464.js" data-settings-js="settings-74424d7eec62a23e.js" ><script src="../static.files/storage-fec3eaa3851e447d.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-5f34af1a0ee6bacd.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-5d8b3c7633ad77ba.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../hydro_deploy/index.html">hydro_deploy</a><span class="version">0.5.0</span></h2></div><div class="sidebar-elems"><ul class="block">
            <li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#reexports">Re-exports</a></li><li><a href="#modules">Modules</a></li><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li><li><a href="#types">Type Aliases</a></li></ul></section></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">hydro_deploy</a><button id="copy-path" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../src/hydro_deploy/lib.rs.html#1-214">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><h2 id="reexports" class="small-section-header"><a href="#reexports">Re-exports</a></h2><ul class="item-table"><li><div class="item-name" id="reexport.Deployment"><code>pub use deployment::<a class="struct" href="deployment/struct.Deployment.html" title="struct hydro_deploy::deployment::Deployment">Deployment</a>;</code></div></li><li><div class="item-name" id="reexport.LocalhostHost"><code>pub use localhost::<a class="struct" href="localhost/struct.LocalhostHost.html" title="struct hydro_deploy::localhost::LocalhostHost">LocalhostHost</a>;</code></div></li><li><div class="item-name" id="reexport.GCPComputeEngineHost"><code>pub use gcp::<a class="struct" href="gcp/struct.GCPComputeEngineHost.html" title="struct hydro_deploy::gcp::GCPComputeEngineHost">GCPComputeEngineHost</a>;</code></div></li><li><div class="item-name" id="reexport.HydroflowCrate"><code>pub use hydroflow_crate::<a class="struct" href="hydroflow_crate/struct.HydroflowCrate.html" title="struct hydro_deploy::hydroflow_crate::HydroflowCrate">HydroflowCrate</a>;</code></div></li><li><div class="item-name" id="reexport.CustomService"><code>pub use custom_service::<a class="struct" href="custom_service/struct.CustomService.html" title="struct hydro_deploy::custom_service::CustomService">CustomService</a>;</code></div></li></ul><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="custom_service/index.html" title="mod hydro_deploy::custom_service">custom_service</a></div></li><li><div class="item-name"><a class="mod" href="deployment/index.html" title="mod hydro_deploy::deployment">deployment</a></div></li><li><div class="item-name"><a class="mod" href="gcp/index.html" title="mod hydro_deploy::gcp">gcp</a></div></li><li><div class="item-name"><a class="mod" href="hydroflow_crate/index.html" title="mod hydro_deploy::hydroflow_crate">hydroflow_crate</a></div></li><li><div class="item-name"><a class="mod" href="localhost/index.html" title="mod hydro_deploy::localhost">localhost</a></div></li><li><div class="item-name"><a class="mod" href="progress/index.html" title="mod hydro_deploy::progress">progress</a></div></li><li><div class="item-name"><a class="mod" href="ssh/index.html" title="mod hydro_deploy::ssh">ssh</a></div></li><li><div class="item-name"><a class="mod" href="terraform/index.html" title="mod hydro_deploy::terraform">terraform</a></div></li><li><div class="item-name"><a class="mod" href="util/index.html" title="mod hydro_deploy::util">util</a></div></li></ul><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.ResourceBatch.html" title="struct hydro_deploy::ResourceBatch">ResourceBatch</a></div></li><li><div class="item-name"><a class="struct" href="struct.ResourcePool.html" title="struct hydro_deploy::ResourcePool">ResourcePool</a></div></li><li><div class="item-name"><a class="struct" href="struct.ResourceResult.html" title="struct hydro_deploy::ResourceResult">ResourceResult</a></div></li></ul><h2 id="enums" class="small-section-header"><a href="#enums">Enums</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.ClientStrategy.html" title="enum hydro_deploy::ClientStrategy">ClientStrategy</a></div><div class="desc docblock-short">Like BindType, but includes metadata for determining whether a connection is possible.</div></li><li><div class="item-name"><a class="enum" href="enum.HostTargetType.html" title="enum hydro_deploy::HostTargetType">HostTargetType</a></div></li><li><div class="item-name"><a class="enum" href="enum.ServerStrategy.html" title="enum hydro_deploy::ServerStrategy">ServerStrategy</a></div><div class="desc docblock-short">Types of connections that a host can make to another host.</div></li></ul><h2 id="traits" class="small-section-header"><a href="#traits">Traits</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.Host.html" title="trait hydro_deploy::Host">Host</a></div></li><li><div class="item-name"><a class="trait" href="trait.LaunchedBinary.html" title="trait hydro_deploy::LaunchedBinary">LaunchedBinary</a></div></li><li><div class="item-name"><a class="trait" href="trait.LaunchedHost.html" title="trait hydro_deploy::LaunchedHost">LaunchedHost</a></div></li><li><div class="item-name"><a class="trait" href="trait.Service.html" title="trait hydro_deploy::Service">Service</a></div></li><li><div class="item-name"><a class="trait" href="trait.ServiceBuilder.html" title="trait hydro_deploy::ServiceBuilder">ServiceBuilder</a></div></li></ul><h2 id="types" class="small-section-header"><a href="#types">Type Aliases</a></h2><ul class="item-table"><li><div class="item-name"><a class="type" href="type.HostStrategyGetter.html" title="type hydro_deploy::HostStrategyGetter">HostStrategyGetter</a></div></li></ul></section></div></main></body></html>