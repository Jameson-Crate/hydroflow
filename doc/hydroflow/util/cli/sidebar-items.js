window.SIDEBAR_ITEMS = {"enum":["BindConfig","BoundConnection","ConnectionPipe"],"fn":["init"],"struct":["ConnectedBidi","ConnectedDemux"],"trait":["Connected"]};