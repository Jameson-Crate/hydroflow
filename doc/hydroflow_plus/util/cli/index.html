<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `cli` mod in crate `hydroflow_plus`."><title>hydroflow_plus::util::cli - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../../static.files/rustdoc-e935ef01ae1c1829.css"><meta name="rustdoc-vars" data-root-path="../../../" data-static-root-path="../../../static.files/" data-current-crate="hydroflow_plus" data-themes="" data-resource-suffix="" data-rustdoc-version="1.78.0-nightly (8f359beca 2024-02-23)" data-channel="nightly" data-search-js="search-dd67cee4cfa65049.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../../static.files/storage-4c98445ec4002617.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../../static.files/main-c37d3936c59ababd.js"></script><noscript><link rel="stylesheet" href="../../../static.files/noscript-04d5337699b92874.css"></noscript><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../../hydroflow_plus/index.html">hydroflow_plus</a><span class="version">0.6.0</span></h2></div><h2 class="location"><a href="#">Module cli</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li><li><a href="#functions">Functions</a></li><li><a href="#types">Type Aliases</a></li></ul></section><h2><a href="../index.html">In hydroflow_plus::util</a></h2></div></nav><div class="sidebar-resizer"></div>
    <main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../../../hydroflow_plus/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../../../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../../index.html">hydroflow_plus</a>::<wbr><a href="../index.html">util</a>::<wbr><a class="mod" href="#">cli</a><button id="copy-path" title="Copy item path to clipboard"><img src="../../../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../../../src/hydroflow/util/mod.rs.html#29">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.ConnectedDemux.html" title="struct hydroflow_plus::util::cli::ConnectedDemux">ConnectedDemux</a></div></li><li><div class="item-name"><a class="struct" href="struct.ConnectedDirect.html" title="struct hydroflow_plus::util::cli::ConnectedDirect">ConnectedDirect</a></div></li><li><div class="item-name"><a class="struct" href="struct.ConnectedTagged.html" title="struct hydroflow_plus::util::cli::ConnectedTagged">ConnectedTagged</a></div></li><li><div class="item-name"><a class="struct" href="struct.DemuxDrain.html" title="struct hydroflow_plus::util::cli::DemuxDrain">DemuxDrain</a></div></li><li><div class="item-name"><a class="struct" href="struct.HydroCLI.html" title="struct hydroflow_plus::util::cli::HydroCLI">HydroCLI</a></div></li><li><div class="item-name"><a class="struct" href="struct.MergeSource.html" title="struct hydroflow_plus::util::cli::MergeSource">MergeSource</a></div></li><li><div class="item-name"><a class="struct" href="struct.TaggedSource.html" title="struct hydroflow_plus::util::cli::TaggedSource">TaggedSource</a></div></li></ul><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.BoundConnection.html" title="enum hydroflow_plus::util::cli::BoundConnection">BoundConnection</a></div></li><li><div class="item-name"><a class="enum" href="enum.RealizedServerPort.html" title="enum hydroflow_plus::util::cli::RealizedServerPort">RealizedServerPort</a></div></li><li><div class="item-name"><a class="enum" href="enum.ServerBindConfig.html" title="enum hydroflow_plus::util::cli::ServerBindConfig">ServerBindConfig</a></div></li><li><div class="item-name"><a class="enum" href="enum.ServerOrBound.html" title="enum hydroflow_plus::util::cli::ServerOrBound">ServerOrBound</a></div></li><li><div class="item-name"><a class="enum" href="enum.ServerPort.html" title="enum hydroflow_plus::util::cli::ServerPort">ServerPort</a></div><div class="desc docblock-short">Describes how to connect to a service which is listening on some port.</div></li></ul><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.Connected.html" title="trait hydroflow_plus::util::cli::Connected">Connected</a></div></li><li><div class="item-name"><a class="trait" href="trait.ConnectedSink.html" title="trait hydroflow_plus::util::cli::ConnectedSink">ConnectedSink</a></div></li><li><div class="item-name"><a class="trait" href="trait.ConnectedSource.html" title="trait hydroflow_plus::util::cli::ConnectedSource">ConnectedSource</a></div></li><li><div class="item-name"><a class="trait" href="trait.StreamSink.html" title="trait hydroflow_plus::util::cli::StreamSink">StreamSink</a></div></li></ul><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="fn" href="fn.init.html" title="fn hydroflow_plus::util::cli::init">init</a></div></li><li><div class="item-name"><a class="fn" href="fn.launch.html" title="fn hydroflow_plus::util::cli::launch">launch</a></div></li><li><div class="item-name"><a class="fn" href="fn.launch_flow.html" title="fn hydroflow_plus::util::cli::launch_flow">launch_flow</a></div></li></ul><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="type" href="type.BufferedDrain.html" title="type hydroflow_plus::util::cli::BufferedDrain">BufferedDrain</a></div></li><li><div class="item-name"><a class="type" href="type.DynSink.html" title="type hydroflow_plus::util::cli::DynSink">DynSink</a></div></li><li><div class="item-name"><a class="type" href="type.DynStream.html" title="type hydroflow_plus::util::cli::DynStream">DynStream</a></div></li><li><div class="item-name"><a class="type" href="type.DynStreamSink.html" title="type hydroflow_plus::util::cli::DynStreamSink">DynStreamSink</a></div></li><li><div class="item-name"><a class="type" href="type.InitConfig.html" title="type hydroflow_plus::util::cli::InitConfig">InitConfig</a></div></li></ul></section></div></main></body></html>