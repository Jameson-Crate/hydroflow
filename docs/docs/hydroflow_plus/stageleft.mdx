---
title: Stageleft
sidebar_position: 4
---

import StageleftDocs from '../../../stageleft/README.md'

<StageleftDocs/>
