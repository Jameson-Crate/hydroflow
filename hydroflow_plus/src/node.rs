use std::cell::RefCell;
use std::rc::Rc;

use hydroflow::util::cli::HydroCLI;
use stageleft::RuntimeData;

pub trait HfNode<'a>: Clone {
    fn id(&self) -> usize;
    fn next_port(&self) -> String;
    fn get_cli(&self) -> RuntimeData<&'a HydroCLI>;
}

impl<'a> HfNode<'a> for () {
    fn id(&self) -> usize {
        0
    }

    fn next_port(&self) -> String {
        panic!();
    }

    fn get_cli(&self) -> RuntimeData<&'a HydroCLI> {
        panic!();
    }
}

impl<'a> HfNode<'a> for usize {
    fn id(&self) -> usize {
        *self
    }

    fn next_port(&self) -> String {
        panic!();
    }

    fn get_cli(&self) -> RuntimeData<&'a HydroCLI> {
        panic!();
    }
}

#[derive(Clone)]
pub struct CLIRuntimeNode<'a> {
    id: usize,
    next_port: Rc<RefCell<usize>>,
    cli: RuntimeData<&'a HydroCLI>,
}

impl<'a> CLIRuntimeNode<'a> {
    pub fn new(id: usize, cli: RuntimeData<&'a HydroCLI>) -> CLIRuntimeNode {
        CLIRuntimeNode {
            id,
            next_port: Rc::new(RefCell::new(0)),
            cli,
        }
    }
}

impl<'a> HfNode<'a> for CLIRuntimeNode<'a> {
    fn id(&self) -> usize {
        self.id
    }

    fn next_port(&self) -> String {
        let next_send_port = *self.next_port.borrow();
        *self.next_port.borrow_mut() += 1;
        format!("port_{}", next_send_port)
    }

    fn get_cli(&self) -> RuntimeData<&'a HydroCLI> {
        self.cli
    }
}

pub trait HfConnectable<'a, O: HfNode<'a>> {
    fn connect(&self, other: &O, source_port: &str, recipient_port: &str);
}

impl<'a> HfConnectable<'a, CLIRuntimeNode<'a>> for CLIRuntimeNode<'a> {
    fn connect(&self, _other: &CLIRuntimeNode, _source_port: &str, _recipient_port: &str) {}
}

pub trait HfNodeBuilder<N> {
    fn build(&mut self, id: usize) -> N;
}

pub struct CLIRuntimeNodeBuilder<'a> {
    cli: RuntimeData<&'a HydroCLI>,
}

impl CLIRuntimeNodeBuilder<'_> {
    pub fn new(cli: RuntimeData<&HydroCLI>) -> CLIRuntimeNodeBuilder {
        CLIRuntimeNodeBuilder { cli }
    }
}

impl<'a> HfNodeBuilder<CLIRuntimeNode<'a>> for CLIRuntimeNodeBuilder<'a> {
    fn build(&mut self, id: usize) -> CLIRuntimeNode<'a> {
        CLIRuntimeNode::new(id, self.cli)
    }
}

pub trait HFDeploy<'a> {
    type Node: HfNode<'a> + HfConnectable<'a, Self::Node>;
    type NodeBuilder: HfNodeBuilder<Self::Node>;
}

pub struct CLIRuntime<'b> {
    _marker: std::marker::PhantomData<&'b ()>,
}

impl<'a: 'b, 'b> HFDeploy<'a> for CLIRuntime<'b> {
    type Node = CLIRuntimeNode<'a>;
    type NodeBuilder = CLIRuntimeNodeBuilder<'a>;
}
