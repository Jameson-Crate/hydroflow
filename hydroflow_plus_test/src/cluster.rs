use hydroflow_plus::node::*;
use hydroflow_plus::*;
use stageleft::*;

pub fn simple_cluster<'a, D: HfNetworkedDeploy<'a>>(
    graph: &'a HfBuilder<'a, D>,
    node_builder: &impl HfNodeBuilder<'a, D>,
    cluster_builder: &impl HfClusterBuilder<'a, D>,
) -> (D::Node, D::Cluster) {
    let node = graph.node(node_builder);
    let cluster = graph.cluster(cluster_builder);

    let numbers = node.source_iter(q!(0..5));
    let ids = node.source_iter(cluster.ids()).map(q!(|&id| id));

    ids.cross_product(&numbers)
        .map(q!(|(id, n)| (id, (id, n))))
        .demux_bincode(&cluster)
        .inspect(q!(|n| println!("cluster received: {:?}", n)))
        .send_bincode_tagged(&node)
        .for_each(q!(|(id, d)| println!("node received: ({}, {:?})", id, d)));

    (node, cluster)
}

pub fn many_to_many<'a, D: HfNetworkedDeploy<'a>>(
    graph: &'a HfBuilder<'a, D>,
    cluster_builder: &impl HfClusterBuilder<'a, D>,
) -> D::Cluster {
    let cluster = graph.cluster(cluster_builder);
    cluster
        .source_iter(q!(0..2))
        .broadcast_bincode_tagged(&cluster)
        .for_each(q!(|n| println!("cluster received: {:?}", n)));

    cluster
}

pub fn map_reduce<'a, D: HfNetworkedDeploy<'a>>(
    graph: &'a HfBuilder<'a, D>,
    node_builder: &impl HfNodeBuilder<'a, D>,
    cluster_builder: &impl HfClusterBuilder<'a, D>,
) -> (D::Node, D::Cluster) {
    let node = graph.node(node_builder);
    let cluster = graph.cluster(cluster_builder);

    let words = node
        .source_iter(q!(vec!["abc", "abc", "xyz"]))
        .map(q!(|s| s.to_string()));

    let all_ids_vec = cluster.ids();
    let words_partitioned = words.enumerate().map(q!({
        let cluster_size = all_ids_vec.len();
        move |(i, w)| ((i % cluster_size) as u32, w)
    }));

    words_partitioned
        .demux_bincode(&cluster)
        .fold(q!(|| 0), q!(|count, string: String| *count += string.len()))
        .inspect(q!(|count| println!("partition count: {}", count)))
        .send_bincode_tagged(&node)
        .persist()
        .map(q!(|(_mid, count)| count))
        .fold(q!(|| 0), q!(|total, count| *total += count))
        .for_each(q!(|data| println!("total: {}", data)));

    (node, cluster)
}

use hydroflow::util::cli::HydroCLI;
use hydroflow_plus_cli_integration::{CLIRuntime, HydroflowPlusMeta};

#[stageleft::entry]
pub fn simple_cluster_runtime<'a>(
    graph: &'a HfBuilder<'a, CLIRuntime>,
    cli: RuntimeData<&'a HydroCLI<HydroflowPlusMeta>>,
    node_id: RuntimeData<usize>,
) -> impl Quoted<'a, Hydroflow<'a>> {
    let _ = simple_cluster(graph, &cli, &cli);
    graph.build(node_id)
}

#[stageleft::entry]
pub fn many_to_many_runtime<'a>(
    graph: &'a HfBuilder<'a, CLIRuntime>,
    cli: RuntimeData<&'a HydroCLI<HydroflowPlusMeta>>,
    node_id: RuntimeData<usize>,
) -> impl Quoted<'a, Hydroflow<'a>> {
    let _ = many_to_many(graph, &cli);
    graph.build(node_id)
}

#[stageleft::entry]
pub fn map_reduce_runtime<'a>(
    graph: &'a HfBuilder<'a, CLIRuntime>,
    cli: RuntimeData<&'a HydroCLI<HydroflowPlusMeta>>,
    node_id: RuntimeData<usize>,
) -> impl Quoted<'a, Hydroflow<'a>> {
    let _ = map_reduce(graph, &cli, &cli);
    graph.build(node_id)
}

#[stageleft::runtime]
#[cfg(test)]
mod tests {
    use std::cell::RefCell;
    use std::time::Duration;

    use hydro_deploy::{Deployment, HydroflowCrate};
    use hydroflow::lattices::cc_traits::Iter;
    use hydroflow_plus_cli_integration::{
        CLIDeployClusterBuilder, CLIDeployNodeBuilder, DeployCrateWrapper,
    };

    #[tokio::test]
    async fn simple_cluster() {
        let deployment = RefCell::new(Deployment::new());
        let localhost = deployment.borrow_mut().Localhost();

        let builder = hydroflow_plus::HfBuilder::new();
        let (node, cluster) = super::simple_cluster(
            &builder,
            &CLIDeployNodeBuilder::new(|id| {
                deployment.borrow_mut().add_service(
                    HydroflowCrate::new(".", localhost.clone())
                        .bin("simple_cluster")
                        .profile("dev")
                        .args(vec![id.to_string()]),
                )
            }),
            &CLIDeployClusterBuilder::new(|id| {
                (0..2)
                    .map(|_| {
                        deployment.borrow_mut().add_service(
                            HydroflowCrate::new(".", localhost.clone())
                                .bin("simple_cluster")
                                .profile("dev")
                                .args(vec![id.to_string()]),
                        )
                    })
                    .collect()
            }),
        );
        builder.wire();

        let mut deployment = deployment.into_inner();

        deployment.deploy().await.unwrap();

        let node_stdout = node.stdout().await;
        let cluster_stdouts =
            futures::future::join_all(cluster.nodes.iter().map(|node| node.stdout())).await;

        deployment.start().await.unwrap();

        for (i, stdout) in cluster_stdouts.into_iter().enumerate() {
            for j in 0..5 {
                assert_eq!(
                    tokio::time::timeout(Duration::from_secs(30), stdout.recv())
                        .await
                        .unwrap()
                        .unwrap(),
                    format!("cluster received: ({}, {})", i, j)
                );
            }
        }

        let mut node_outs = vec![];
        for _i in 0..10 {
            node_outs.push(
                tokio::time::timeout(Duration::from_secs(30), node_stdout.recv())
                    .await
                    .unwrap()
                    .unwrap(),
            );
        }
        node_outs.sort();

        for (i, n) in node_outs.into_iter().enumerate() {
            assert_eq!(
                n,
                format!("node received: ({}, ({}, {}))", i / 5, i / 5, i % 5)
            );
        }
    }

    #[tokio::test]
    async fn many_to_many() {
        let deployment = RefCell::new(Deployment::new());
        let localhost = deployment.borrow_mut().Localhost();

        let builder = hydroflow_plus::HfBuilder::new();
        let cluster = super::many_to_many(
            &builder,
            &CLIDeployClusterBuilder::new(|id| {
                (0..2)
                    .map(|_| {
                        deployment.borrow_mut().add_service(
                            HydroflowCrate::new(".", localhost.clone())
                                .bin("many_to_many")
                                .profile("dev")
                                .args(vec![id.to_string()]),
                        )
                    })
                    .collect()
            }),
        );
        builder.wire();

        let mut deployment = deployment.into_inner();

        deployment.deploy().await.unwrap();

        let cluster_stdouts =
            futures::future::join_all(cluster.nodes.iter().map(|node| node.stdout())).await;

        deployment.start().await.unwrap();

        for node_stdout in cluster_stdouts {
            let mut node_outs = vec![];
            for _i in 0..4 {
                node_outs.push(
                    tokio::time::timeout(Duration::from_secs(30), node_stdout.recv())
                        .await
                        .unwrap()
                        .unwrap(),
                );
            }
            node_outs.sort();

            let mut node_outs = node_outs.into_iter();

            for sender in 0..2 {
                for value in 0..2 {
                    assert_eq!(
                        node_outs.next().unwrap(),
                        format!("cluster received: ({}, {})", sender, value)
                    );
                }
            }
        }
    }
}
