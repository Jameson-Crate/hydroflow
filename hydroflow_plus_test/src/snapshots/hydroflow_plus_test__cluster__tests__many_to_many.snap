---
source: hydroflow_plus_test/src/cluster.rs
expression: builder.ir()
---
[
    ForEach {
        f: { use crate :: __staged :: cluster :: * ; | n | println ! ("cluster received: {:?}" , n) },
        input: Network {
            to_location: 0,
            serialize_pipeline: Some(
                Operator(
                    Operator {
                        path: "map",
                        args: [
                            "| (id , data) | { (id , hydroflow_plus :: runtime_support :: bincode :: serialize :: < i32 > (& data) . unwrap () . into ()) }",
                        ],
                    },
                ),
            ),
            sink_expr: null,
            source_expr: null,
            deserialize_pipeline: Some(
                Operator(
                    Operator {
                        path: "map",
                        args: [
                            "| res | { let (id , b) = res . unwrap () ; (id , hydroflow_plus :: runtime_support :: bincode :: deserialize :: < i32 > (& b) . unwrap ()) }",
                        ],
                    },
                ),
            ),
            input: CrossProduct(
                Persist(
                    PipelineOp {
                        kind: "cloned",
                        gen_pipeline: <function>,
                        input: Source {
                            source: Iter(
                                { use hydroflow_plus_cli_integration :: __staged :: deploy :: * ; panic ! () },
                            ),
                            produces_delta: false,
                            location_id: 0,
                        },
                    },
                ),
                Source {
                    source: Iter(
                        { use crate :: __staged :: cluster :: * ; 0 .. 2 },
                    ),
                    produces_delta: false,
                    location_id: 0,
                },
            ),
        },
    },
]
